Updateable layout_provider/graph_layout

g-parki opened the pull request:
- [x] issues: fixes #11474
- [x] tests added / passed

Edit after overhauling:

This PR is to fix the original issue of the graph not updating after the `layout_provider` was updated, and also add transform classes for edge/node coordinates:

In `layout_provider.ts`:
 - Base class `GraphCoordinates` extends `CoordinateTransform`, adds `layout` property
 - Classes `NodeCoordinates` and `EdgeCoordinates` extend `GraphCoordinates`, adds `_v_compute()` method
 - `layout_provider` has getters for `node_coordinates` and `edge_coordinates`

In `coordinate_transform.ts`:
 - CoordinateTransform component types broadened from `Float64Array` to `Arrayable<number> | Arrayable<number>[]`



Screen capture of the graph updating:
![graph_render](https://user-images.githubusercontent.com/61096711/131612670-ffb2e3dc-8479-4d34-bf77-6eb1e7293244.gif)

For reference, here's the script I used for prototyping (based on issue #11474):
```
import networkx as nx
from bokeh.plotting import figure, show, from_networkx
from bokeh.layouts import layout
from bokeh.models import ColumnDataSource, CustomJS, Select, LabelSet

def get_node_label_source(graph, pos):
    renderer = from_networkx(graph, pos)
    node_names = list(renderer.layout_provider.graph_layout.keys())
    _xs, _ys = map(list, zip(*renderer.layout_provider.graph_layout.values()))
    return ColumnDataSource(
        data=dict(
            xs=_xs,
            ys=_ys,
            names=node_names
        )
    )

#define the graph's elements:
source = ['C1', 'C1', 'C1','C3','C2','C2','C2','C4','C5']
target = ['C3', 'C2','C4','C4','C3','C4','C1','C3', 'C3']
weight = [1,0.3,1,1,1,0.8, 0.4, 1, 1]
edges_list = list(zip(source, target, weight))

nx_graph = nx.DiGraph()
nx_graph.add_weighted_edges_from(edges_list)
plot = figure()
node_initial_pos = {'C1':(0, 0), 'C2':(0, 3), 'C5':(3,0)}

_layout_options = ['circular_pos', 'kamada_kawai_pos', 'planar_pos']
circular_pos = nx.circular_layout(nx_graph,scale=2,center=(0, 0),dim=2,)
kamada_kawai_pos = nx.kamada_kawai_layout(nx_graph,weight='weight',scale=2,center=(0, 0),dim=2,)
planar_pos = nx.planar_layout(nx_graph,scale=2,center=(0, 0),dim=2)

initial_graph_renderer = from_networkx(nx_graph, planar_pos)
plot.renderers.append(initial_graph_renderer)

initial_node_labels = LabelSet(
    x='xs',
    y='ys',
    text='names',
    source=get_node_label_source(nx_graph, planar_pos),
)
plot.add_layout(initial_node_labels)

nodes_display_layout_select = Select(
    title="Display layout:",
    value="planar_layout",
    options=_layout_options,
    width=300,
    height=50
)

nodes_display_layout_select_args = dict(
    graph_renderer=initial_graph_renderer,
    plot=plot,
    positions = dict(
        circular_pos=dict(position=circular_pos, label_source= get_node_label_source(nx_graph, circular_pos)),
        kamada_kawai_pos=dict(position=kamada_kawai_pos, label_source= get_node_label_source(nx_graph, kamada_kawai_pos)),
        planar_pos=dict(position=planar_pos, label_source= get_node_label_source(nx_graph, planar_pos))
    ),
    node_labels= initial_node_labels
)

nodes_display_layout_select_str = """
    const selected_layout = positions[cb_obj.value];
    graph_renderer.layout_provider.graph_layout = selected_layout.position;
    node_labels.source= selected_layout.label_source;
"""
nodes_display_layout_select_callback = CustomJS(
    args=nodes_display_layout_select_args,
    code=nodes_display_layout_select_str,
)
nodes_display_layout_select.js_on_change("value", nodes_display_layout_select_callback)

html_layout = layout(plot,nodes_display_layout_select)
show(html_layout, browser="windows-default")
```

mattpap commented on bokehjs/test/unit/models/renderers/graph_renderer.ts:
```diff
@@ -9,34 +9,54 @@ import {Plot} from "@bokehjs/models/plots"
+    const x = node_glyph.x as any
+    const y = node_glyph.y as any
+    const express_x = x.expr as Expression
+    const express_y = y.expr as Expression
```
Much better would be:
```ts
import {assert} from "core/util/assert"
import {isExpr} from "core/vectorization"
assert(isExpr(node_glyph.x))
```
After the `assert(...)` the type of `x` will be an `Expr`.

mattpap commented on bokehjs/test/unit/models/renderers/graph_renderer.ts:
```diff
@@ -9,34 +9,54 @@ import {Plot} from "@bokehjs/models/plots"
+  data_source: new ColumnDataSource({data: {start: [4, 4, 5, 6], end: [5, 6, 6, 7]}}),
+})
+
+const graph = new GraphRenderer({layout_provider, node_renderer, edge_renderer})
```
This code needs to be wrapped in a function and fresh objects need to be produced for every test. This way we can make sure that tests can be run e.g. out-of-order, and there will be no mutable state creeping between tests.

mattpap commented:
I would like a visual regression test to be added, similar to:
https://github.com/bokeh/bokeh/blob/d3fce050d3079541cc1ace704ce80c8e2ad199a1/bokehjs/test/integration/regressions.ts#L51-L66
Test plot area should be as small as possible, but allowing visual inspection, and the data simple, but covering the issue.

g-parki commented:
In `graph_renderer` I noticed there was a TODO to upgrade node/edge coordinates from an anonymous class to a transform. I was already having a typecasting nightmare with my unit tests, so figured it's a good time for the upgrade. Here's the main summary:

- In `layout_provider.ts`:
  - Classes `NodeCoordinates` and `EdgeCoordinates` extend `CoordinateTransform`
  - `layout_provider` has non-generic getters for each
- In `coordinate_transform.ts`:
  -  CoordinateTransform type broadened from `Expression<{x: Float64Array, y: Float64Array}>` to `Expression<{x: CoordinateType, y: CoordinateType}>` where `CoordinateType = Arrayable<number> | Arrayable<number>[]`

g-parki commented:
Running list still to do (assuming this direction is okay)...

1. Implement same models in Bokeh (currently only in JS). This was caught in ci/js-test.
2. Address failing macos tests.

g-parki commented on bokehjs/test/unit/models/graphs/graph_hit_test_policy.ts:
```diff
@@ -15,20 +15,26 @@ import {ColumnarDataSource} from "@bokehjs/models/sources/columnar_data_source"
 
-  get_node_coordinates(graph_source: ColumnarDataSource): [Arrayable<number>, Arrayable<number>] {
+  constructor(attrs?: Partial<LayoutProvider.Attrs>) {
+    super(attrs)
+  }
+
+  get_node_coordinates(graph_source: ColumnarDataSource): [Float64Array, Float64Array] {
     const n = graph_source.get_length() ?? 1
-    return [new Float64Array(n), new Float64Array(n)]
+    const x = new Float64Array(n)
+    const y = new Float64Array(n)
+    return [x, y]
```
Didn't mean to keep these changes. Will remove.

g-parki commented:
@bryevdv sorry I didn't realize 2.4 is due tomorrow. I can bring this PR back to minimally fixing #11474 if the rest is too much to consider by then...

bryevdv commented:
@g-parki my guess is that the release will slip until Monday, there some loose ends that still need cleaning up. But it's fairly easy now for us to follow on with small point releases so I would say there is no need to rush or hurry, whatever pace works for you is fine.

mattpap commented on bokeh/models/graphs.py:
```diff
@@ -57,7 +62,13 @@ class LayoutProvider(Model):
+
+    @property
+    def edge_coordinates(self):
+        return EdgeCoordinates(layout= self)
```
Whitespace after `=` should be removed.

mattpap replied:
New code should have explicit (return) type annotations (even if trivial).

mattpap commented on bokeh/models/graphs.py:
```diff
@@ -78,6 +89,30 @@ class StaticLayoutProvider(LayoutProvider):
+@abstract
+class GraphCoordinates(CoordinateTransform):
+    '''
+    Abstract class for coordinate transform expression obtained from `LayoutProvider`
```
Sphinx requires double backticks for code blocks.

mattpap commented on bokehjs/src/lib/models/graphs/layout_provider.ts:
```diff
@@ -21,4 +22,75 @@ export abstract class LayoutProvider extends Model {
+export abstract class GraphCoordinates extends CoordinateTransform {
+  override properties: GraphCoordinates.Props
+
+  constructor(attrs?: Partial<GraphCoordinates.Attrs>){
```
A single space is required before opening a block. There are multiple other instances to fix. I will add a lint rule for this.

mattpap commented:
@g-parki, looks good. Please rebase and fix the trivia and this should be ready for merging.

g-parki commented:
> @g-parki, looks good. Please rebase and fix the trivia and this should be ready for merging.

@mattpap thanks for reviewing. I rebased and incorporated your suggestions.
